eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java

```diff
@@ -0,0 +1,251 @@
+		this.preExistingTestFile.create(generateFileStreamLines(25), true, null);
+		this.preExistingMarker = this.preExistingTestFile.createMarker(IMarker.PROBLEM);
+		this.preExistingTestFile.createMarker(IMarker.PROBLEM);
+
+		this.workspace.save(true, null);
+
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+	}
+
+	private InputStream generateFileStreamLines(int numLines) {
+		StringBuilder builder = new StringBuilder();
+
+		for (int i = 0; i < numLines; i++) {
+			builder.append("Line ")
+				   .append(i)
+				   .append('\n');
+		}
+
+		return new ByteArrayInputStream(builder.toString().getBytes(StandardCharsets.UTF_8));
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		this.project.delete(true, true, null);
+		this.workspace.save(true, null);
+
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+		this.workbenchListener.shutDown();
+	}
+
+	@Test
+	public void picks_up_created_warnings_on_first_save() throws Exception {
+		IMarker marker = this.testFile.createMarker(IMarker.PROBLEM);
+		marker.setAttribute(IMarker.LINE_NUMBER, 1);
+		IMarker marker2 = this.testFile.createMarker(IMarker.PROBLEM);
+		marker2.setAttribute(IMarker.LINE_NUMBER, 2);
+		this.workspace.save(true, null);
+
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+
+		ArgumentCaptor<StaticAnalysisWarningEvent> captor = ArgumentCaptor.forClass(StaticAnalysisWarningEvent.class);
+		Mockito.verify(this.trackingEventManager, Mockito.times(2)).addEvent(captor.capture());
+		assertTrue(captor.getAllValues().stream()
+				.allMatch(event -> event.getType() == TrackingEventType.SA_WARNING_CREATED));
+	}
+
+	@Test
+	public void does_not_generate_events_for_same_markers() throws Exception {
+		IMarker marker = this.testFile.createMarker(IMarker.PROBLEM);
+		marker.setAttribute(IMarker.LINE_NUMBER, 1);
+		IMarker marker2 = this.testFile.createMarker(IMarker.PROBLEM);
+		marker2.setAttribute(IMarker.LINE_NUMBER, 2);
+
+		this.workspace.save(true, null);
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+
+		this.testFile.appendContents(generateFileStreamLines(10), true, true, null);
+
+		this.workspace.save(true, null);
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+
+		ArgumentCaptor<StaticAnalysisWarningEvent> captor = ArgumentCaptor.forClass(StaticAnalysisWarningEvent.class);
+		Mockito.verify(this.trackingEventManager, Mockito.times(2)).addEvent(captor.capture());
+		assertTrue(captor.getAllValues().stream()
+				.allMatch(event -> event.getType() == TrackingEventType.SA_WARNING_CREATED));
+	}
+
+	@Test
+	public void generates_removal_after_marker_is_deleted() throws Exception {
+		IMarker marker = this.testFile.createMarker(IMarker.PROBLEM);
+		marker.setAttribute(IMarker.LINE_NUMBER, 1);
+		IMarker marker2 = this.testFile.createMarker(IMarker.PROBLEM);
+		marker2.setAttribute(IMarker.LINE_NUMBER, 2);
+
+		this.workspace.save(true, null);
```

Comment:
What is the general rule for this. in the rest of the project? I don't think we usually use it outside constructors, right?

Reply:
I am not sure what you mean. What do you mean with "outside constructors"?